lock/src/main/java/org/infinispan/lock/api/impl/lock/ClusteredLockImpl.java

```diff
@@ -0,0 +1,369 @@
+            // There has to be at least 2 members
+            // FIXME: Rentrant locks, can the node rejoin and reacquire the lock before the release actually happens ?? Use the requestId
+            oldMembers.stream().filter(a -> !newMembers.contains(a)).forEach(notPresent -> {
+               try {
+                  if (clusteredLockManager.isDefined(name)) {
+                     clusteredLockManager.execute(() -> unlock(null, notPresent)
+                           .exceptionally(ex -> {
+                              log.error("Unlock failed wrong", ex);
+                              return null;
+                           }));
+                  }
+               } catch (AvailabilityException ex) {
+                  log.error("Unable to release due to cluster change", ex);
+               }
+            });
+         }
+      }
+   }
+
+   @Override
+   public CompletableFuture<Void> lock() {
+      log.tracef("lock called from %s", originator);
+      CompletableFuture<Void> lockRequest = new CompletableFuture<>();
+      lock(new LockRequestHolder(originator, lockRequest));
+      return lockRequest;
+   }
+
+   private void lock(RequestHolder<Void> requestHolder) {
+      if (requestHolder == null || requestHolder.isDone())
+         return;
+
+      pendingRequests.offer(requestHolder);
+      readWriteMap.eval(lockKey, new LockFunction(requestHolder.requestId, requestHolder.requestor))
+            .whenComplete((lockResult, ex) -> requestHolder.handleLockResult(lockResult, ex));
+   }
+
+   @Override
+   public CompletableFuture<Boolean> tryLock() {
+      log.tracef("tryLock called from %s", originator);
+      CompletableFuture<Boolean> tryLockRequest = new CompletableFuture<>();
+      tryLock(new TryLockRequestHolder(originator, tryLockRequest));
+      return tryLockRequest;
+   }
+
+   @Override
+   public CompletableFuture<Boolean> tryLock(long time, TimeUnit unit) {
+      log.tracef("tryLock with timeout (%l, %s) called from %s", time, unit, originator);
+      CompletableFuture<Boolean> tryLockRequest = new CompletableFuture<>();
+      tryLock(new TryLockRequestHolder(originator, tryLockRequest, time, unit));
+      return tryLockRequest;
+   }
+
+   private void tryLock(TryLockRequestHolder requestHolder) {
+      if (requestHolder == null || requestHolder.isDone()) {
+         return;
+      }
+      if (requestHolder.hasTimeout()) pendingRequests.offer(requestHolder);
+
+      readWriteMap.eval(lockKey, new LockFunction(requestHolder.requestId, requestHolder.requestor)).whenComplete((lockResult, ex) -> {
+         requestHolder.handleLockResult(lockResult, ex);
+      });
+   }
+
+   @Override
+   public CompletableFuture<Void> unlock() {
+      log.tracef("unlock called from %s", originator);
+      CompletableFuture<Void> unlockRequest = new CompletableFuture<>();
+      readWriteMap.eval(lockKey, new UnlockFunction(originator)).whenComplete((lockResult, ex) -> {
```

Comment:
This is not fluent, capturing another future... could you use `.exceptionally(t -> throw handleException(t))` instead of `.whenComplete` (maybe with `CompletionException` wrapping)?